lib/imagec/download.go

```diff
@@ -0,0 +1,330 @@
+// Copyright 2016 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissi[ons and
+// limitations under the License.
+
+package imagec
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	log "github.com/Sirupsen/logrus"
+
+	"github.com/vmware/vic/lib/apiservers/engine/backends/cache"
+
+	"github.com/docker/docker/distribution/xfer"
+	"github.com/docker/docker/pkg/progress"
+)
+
+var ()
+
+const (
+	maxDownloadAttempts    = 5
+	maxConcurrentDownloads = 3
+)
+
+type downloadTransfer struct {
+	xfer.Transfer
+
+	layer *ImageWithMeta
+	err   error
+}
+
+// result returns the layer resulting from the download, if the download
+// and registration were successful.
+func (d *downloadTransfer) result() error {
+	return d.err
+}
+
+// LayerDownloader keeps track of what layers are currently being downloaded
+type LayerDownloader struct {
+	tm xfer.TransferManager
+	// the list of layers currently in flight
+	downloadsByID map[string]*downloadTransfer
+
+	m sync.Mutex
+}
+
+// NewLayerDownloader creates and returns a new LayerDownloadManager
+func NewLayerDownloader() *LayerDownloader {
+	return &LayerDownloader{
+		tm:            xfer.NewTransferManager(maxConcurrentDownloads),
+		downloadsByID: make(map[string]*downloadTransfer),
+	}
+}
+
+func (ldm *LayerDownloader) registerDownload(download *downloadTransfer) {
+	ldm.downloadsByID[download.layer.ID] = download
+}
+
+func (ldm *LayerDownloader) unregisterDownload(id string) {
+	// mark the layer as finished downloading
+	cache.LayerCache().Commit(id)
+
+	// stop tracking the download transfer
+	delete(ldm.downloadsByID, id)
+}
+
+// DownloadLayers ensures layers end up in the portlayer's image store
+// It handles existing and simultaneous layer download de-duplication
+// This code is utilizes Docker's xfer package: https://github.com/docker/docker/tree/v1.11.2/distribution/xfer
+func (ldm *LayerDownloader) DownloadLayers(ctx context.Context, ic *ImageC) error {
```

Comment:
@hmahmood and I were asking ourselves the same question and it comes down to implementing some of Docker's interfaces. I believe we can port our download code over to these interface implementations, but there are a few things I'm not sure of, like where the line is between our layers and theirs (they call a lot of functions that are defined on docker layers that we don't have). I will look for a clear line of separation in order to make this possible. It would go a long way to making things a lot cleaner and easier to understand.